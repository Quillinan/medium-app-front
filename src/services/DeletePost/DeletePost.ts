import { apiUrl, deleteApi } from '@services/Api/Api';
import handleError from '@errors/HandleError/HandleError';
import { ErrorResponse } from '@utils/Types/Types';

export const deletePost = async (
  authorId: string
): Promise<void | ErrorResponse> => {
  const url = `${apiUrl}api/Post/${authorId}`;

  try {
    const result = await deleteApi(url);

    if (result === undefined) {
      return;
    }

    return result as ErrorResponse;
  } catch (error: any) {
    return handleError(error);
  }
};
