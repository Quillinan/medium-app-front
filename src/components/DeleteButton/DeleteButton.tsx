import { deletePost } from '@services/DeletePost/DeletePost';
import React from 'react';
import Swal from 'sweetalert2';
import { useNavigate } from 'react-router-dom';

interface DeleteButtonProps {
  postId: string;
  onDeleteSuccess: () => void;
}

const DeleteButton: React.FC<DeleteButtonProps> = ({
  postId,
  onDeleteSuccess,
}) => {
  const navigate = useNavigate();

  const handleDelete = async () => {
    const confirmation = await Swal.fire({
      title: 'Tem certeza?',
      text: 'Você realmente quer deletar este post?',
      icon: 'warning',
      showCancelButton: true,
      confirmButtonText: 'Sim, deletar',
      cancelButtonText: 'Cancelar',
    });

    if (confirmation.isConfirmed) {
      const result = await deletePost(postId);

      if (!result) {
        Swal.fire({
          title: 'Sucesso',
          text: 'Post deletado com sucesso!',
          icon: 'success',
        }).then(() => {
          onDeleteSuccess();
          navigate('/');
        });
      } else {
        console.error('Erro ao excluir o post:', result);
        Swal.fire({
          title: 'Erro',
          text: result.message || 'Não foi possível excluir o post.',
          icon: 'error',
        });
      }
    }
  };

  return (
    <button
      onClick={handleDelete}
      className='bg-red-500 text-white rounded px-4 py-1'
    >
      Apagar
    </button>
  );
};

export default DeleteButton;
