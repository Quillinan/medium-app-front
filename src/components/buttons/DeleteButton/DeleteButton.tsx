import { deletePost } from '@services/DeletePost/DeletePost';
import React from 'react';
import Swal from 'sweetalert2';

interface DeleteButtonProps {
  postId: string;
  onDeleteSuccess: () => void;
}

const DeleteButton: React.FC<DeleteButtonProps> = ({
  postId,
  onDeleteSuccess,
}) => {
  const handleDelete = async () => {
    const confirmation = await Swal.fire({
      title: 'Tem certeza?',
      text: 'Você realmente quer deletar este post?',
      icon: 'warning',
      showCancelButton: true,
      confirmButtonText: 'Sim, deletar',
      cancelButtonText: 'Cancelar',
    });

    if (confirmation.isConfirmed) {
      const result = await deletePost(postId);

      if (!result) {
        Swal.fire({
          title: 'Sucesso',
          text: 'Post deletado com sucesso!',
          icon: 'success',
        }).then(() => {
          onDeleteSuccess();
          window.location.reload();
        });
      } else {
        console.error('Erro ao excluir o post:', result);
        Swal.fire({
          title: 'Erro',
          text: result.message || 'Não foi possível excluir o post.',
          icon: 'error',
        });
      }
    }
  };

  return (
    <button
      onClick={handleDelete}
      className='bg-red-500 text-white rounded px-4 py-1'
    >
      Apagar
    </button>
  );
};

export default DeleteButton;
