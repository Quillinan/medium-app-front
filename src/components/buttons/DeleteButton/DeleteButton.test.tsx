import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import DeleteButton from './DeleteButton';
import { deletePost } from '@services/DeletePost/DeletePost';
import Swal from 'sweetalert2';
import { Mock } from 'vitest';

vi.mock('@services/DeletePost/DeletePost');
vi.mock('sweetalert2');

describe('DeleteButton', () => {
  const mockOnDeleteSuccess = vi.fn();

  beforeEach(() => {
    mockOnDeleteSuccess.mockClear();
    (Swal.fire as Mock).mockClear();
    (deletePost as Mock).mockClear();
    (Swal.fire as Mock).mockResolvedValue({ isConfirmed: true });
  });

  beforeAll(() => {
    vi.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterAll(() => {
    (console.error as Mock).mockRestore();
  });

  it('should not call deletePost if cancellation is confirmed', async () => {
    (Swal.fire as Mock).mockResolvedValueOnce({ isConfirmed: false });

    render(<DeleteButton postId='1' onDeleteSuccess={mockOnDeleteSuccess} />);

    fireEvent.click(screen.getByText(/Apagar/i));

    await waitFor(() =>
      expect(Swal.fire).toHaveBeenCalledWith(
        expect.objectContaining({
          title: 'Tem certeza?',
          text: 'Você realmente quer deletar este post?',
        })
      )
    );

    expect(deletePost).not.toHaveBeenCalled();
    expect(mockOnDeleteSuccess).not.toHaveBeenCalled();
  });

  it('should show error alert if deletePost fails', async () => {
    (deletePost as Mock).mockResolvedValue({ message: 'Erro de rede' });

    render(<DeleteButton postId='1' onDeleteSuccess={mockOnDeleteSuccess} />);

    fireEvent.click(screen.getByText(/Apagar/i));

    await waitFor(() =>
      expect(Swal.fire).toHaveBeenCalledWith(
        expect.objectContaining({
          title: 'Tem certeza?',
          text: 'Você realmente quer deletar este post?',
        })
      )
    );

    (Swal.fire as Mock).mockResolvedValueOnce({ isConfirmed: true });

    await waitFor(() => expect(deletePost).toHaveBeenCalledWith('1'));
    expect(Swal.fire).toHaveBeenCalledWith({
      title: 'Erro',
      text: 'Erro de rede',
      icon: 'error',
    });
    expect(mockOnDeleteSuccess).not.toHaveBeenCalled();
  });

  it('should show success alert and call onDeleteSuccess when post is deleted', async () => {
    (deletePost as Mock).mockResolvedValue(null);

    render(<DeleteButton postId='1' onDeleteSuccess={mockOnDeleteSuccess} />);

    fireEvent.click(screen.getByText(/Apagar/i));

    await waitFor(() =>
      expect(Swal.fire).toHaveBeenCalledWith(
        expect.objectContaining({
          title: 'Tem certeza?',
          text: 'Você realmente quer deletar este post?',
        })
      )
    );

    (Swal.fire as Mock).mockResolvedValueOnce({ isConfirmed: true });

    await waitFor(() => expect(deletePost).toHaveBeenCalledWith('1'));
    expect(Swal.fire).toHaveBeenCalledWith({
      title: 'Sucesso',
      text: 'Post deletado com sucesso!',
      icon: 'success',
    });
    expect(mockOnDeleteSuccess).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <DeleteButton postId='1' onDeleteSuccess={mockOnDeleteSuccess} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
