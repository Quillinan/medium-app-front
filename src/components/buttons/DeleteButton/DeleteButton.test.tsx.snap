// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeleteButton > should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="bg-red-500 text-white rounded px-4 py-1"
  >
    Apagar
  </button>
</DocumentFragment>
`;
